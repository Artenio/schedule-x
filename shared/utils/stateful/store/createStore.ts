export default function createStore<T extends object>(data: T, name: string) {
  function emit(type: string, detail: object) {
    const event = new CustomEvent(type, {
      bubbles: true,
      cancelable: true,
      detail: detail,
    })

    return document.dispatchEvent(event)
  }

  return new Proxy<T>(data, {
    get: function (obj, prop) {
      return obj[prop as keyof typeof obj]
    },
    set: function (obj: T, prop: string, value: unknown) {
      console.log('set ' + prop + ' to ' + value + ' in ' + name)

      const propToSet = prop as keyof T

      if (obj[propToSet as keyof T] === value) return true

      obj[propToSet] = value as T[keyof T]
      emit(name, data)

      return true
    },
    deleteProperty: function (obj, prop) {
      delete obj[prop as keyof typeof obj]
      emit(name, data)

      return true
    },
  })
}
