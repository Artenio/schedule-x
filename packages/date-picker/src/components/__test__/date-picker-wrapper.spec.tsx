import {
  describe,
  it,
  expect,
  beforeEach,
} from '../../../../../shared/utils/stateless/testing/unit/unit-testing-library.impl'
import { render, fireEvent, screen, waitFor } from '@testing-library/preact'
import DatePickerWrapper from '../date-picker-wrapper'
import createStore from '../../../../../shared/utils/stateful/store/createStore'
import DatePickerStateBuilder from '../../../../../shared/utils/stateful/date-picker-state/date-picker-state.builder'
import TimeUnitsBuilder from '../../../../../shared/utils/stateful/time-units/time-units.builder'
import { StoreModuleName } from '../../../../../shared/enums/store-module-name.enum'
import DatePickerSingleton from '../../interfaces/app-singleton.interface'

describe('date picker wrapper', () => {
  let $app: DatePickerSingleton | undefined

  beforeEach(() => {
    $app = {
      datePickerState: createStore(
        new DatePickerStateBuilder().build(),
        'DATE_PICKER_STATE'
      ),
      timeUnitsImpl: new TimeUnitsBuilder().build(),
    }
  })

  it('should not display popup on render', () => {
    const { container } = render(<DatePickerWrapper $app={$app!} config={{}} />)

    expect(screen.queryByTestId('date-picker-popup')).toBeNull()
  })

  it('should display popup on render', () => {
    $app!.datePickerState.open()
    const { container } = render(<DatePickerWrapper $app={$app!} config={{}} />)

    expect(screen.queryByTestId('date-picker-popup')).not.toBeNull()
  })

  it('should display popup when datePickerState changes and isOpen is true', async () => {
    const { container } = render(<DatePickerWrapper $app={$app!} config={{}} />)

    const customEvent = new CustomEvent(StoreModuleName.DATE_PICKER_STATE, {
      detail: { isOpen: true },
    })
    document.dispatchEvent(customEvent)

    await waitFor(() => {
      expect(screen.queryByTestId('date-picker-popup')).not.toBeNull()
    })
  })
})
