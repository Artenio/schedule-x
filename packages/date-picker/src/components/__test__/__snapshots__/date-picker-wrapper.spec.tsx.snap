// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`date picker wrapper > should not display popup on render 1`] = `
<div>
  <div>
    <div>
      <input
        data-testid="date-picker-input"
        type="text"
      />
    </div>
  </div>
</div>
`;
